### PR TITLE
TW-902: new design group informations

### PR DESCRIPTION
**Pull request by imGok**
### Ticket
- #902

#### Mobile

https://github.com/linagora/twake-on-matrix/assets/99852347/ae6980e7-c463-4f64-90b9-eaa20e11bbfa

#### WEB

1. Only update avatar success

https://github.com/linagora/twake-on-matrix/assets/99852347/e8875564-ba1d-4655-9e74-7f71ac498e4f

2. Only update group name success

https://github.com/linagora/twake-on-matrix/assets/99852347/e2c75f56-35c7-49ad-8166-35ec2820471c

3. Only update group descriptions success
 
https://github.com/linagora/twake-on-matrix/assets/99852347/85922614-41e7-454a-99e0-01e3a2ae2167

4. Update all success

https://github.com/linagora/twake-on-matrix/assets/99852347/0e6f64d5-168c-4289-929c-d2865cd9d10d

5. Only update group avatar failure

https://github.com/linagora/twake-on-matrix/assets/99852347/2bdb6bd8-5a81-455c-9761-5476ba8faaa3

6. Update group avatar success and group name is failure

https://github.com/linagora/twake-on-matrix/assets/99852347/bddee7e7-f801-4487-a3fa-06f6cf5af74d

7. Update group avatar success and group name success, des is failure

https://github.com/linagora/twake-on-matrix/assets/99852347/76e7a5de-c47f-4b36-a416-d4f05b8f4231

### CODE REVIEW

File: lib/domain/usecase/room/update_group_chat_interactor.dart [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        await room.setName(
+          displayName,
+        );
+      }
+
+      if (description != null) {
+        await room.setDescription(
+          description,
+        );
+      }
```
Comment (hoangdat) [inline lib/domain/usecase/room/update_group_chat_interactor.dart:39]: This is separated API? How we can do it in a single API request?

File: lib/domain/usecase/room/update_group_chat_interactor.dart [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+      yield Right(
+        UpdateGroupChatSuccess(
+          displayName: displayName,
+          roomAvatar: avatar,
+          description: description,
+          isDeleteAvatar: isDeleteAvatar,
+        ),
```
Comment (hoangdat) [inline lib/domain/usecase/room/update_group_chat_interactor.dart:47]: what happen if one of these actions is failed?

Comment (hoangdat) [reply lib/domain/usecase/room/update_group_chat_interactor.dart:47]: why you need to push back these info when success?

File: lib/pages/chat/chat.dart [hunk 3]
```diff
@@ -1584,18 +1583,11 @@ class ChatController extends State<Chat>
...
-      return;
-    }
-    final result = await context.push('/rooms/${room!.id}/details');
-    if (result is ChatDetailsActions) {
-      switch (result) {
-        case ChatDetailsActions.search:
-          toggleSearch();
-          break;
-        default:
-      }
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:1598]: dont forget this case.

**Comment by hoangdat** [comment]
the hight of appBar is not similar

image

File: lib/domain/usecase/room/update_group_chat_interactor.dart [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import 'package:fluffychat/app_state/success.dart';
+import 'package:fluffychat/domain/app_state/room/update_group_chat_failure.dart';
+import 'package:fluffychat/domain/app_state/room/update_group_chat_loading.dart';
+import 'package:fluffychat/domain/app_state/room/update_group_chat_success.dart';
+import 'package:matrix/matrix.dart';
+
+class UpdateGroupChatInteractor {
+  Stream<Either<Failure, Success>> execute({
+    required Room room,
+    MatrixFile? avatar,
```
Comment (hoangdat) [inline lib/domain/usecase/room/update_group_chat_interactor.dart:12]: handle the case of avatar

File: assets/l10n/intl_en.arb [hunk 5]
```diff
@@ -2955,5 +2955,11 @@
   "enterCaption": "Enter a caption",
   "failToSend": "Failed to send, please try again",
   "showLess": "Show Less",
-  "showMore": "Show More"
+  "showMore": "Show More",
+  "groupInformations": "Group informations",
```
Comment (sherlockvn) [inline assets/l10n/intl_en.arb:2959]: information don't have plural right ?

File: lib/domain/app_state/room/update_group_chat_success.dart [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+class UpdateGroupChatSuccess extends Success {
+  final MatrixFile? roomAvatar;
+  final String? displayName;
+  final String? description;
+  final bool isDeleteAvatar;
+
+  const UpdateGroupChatSuccess({
+    this.roomAvatar,
```
Comment (sherlockvn) [inline lib/domain/app_state/room/update_group_chat_success.dart:16]: confuse name, to `roomAvatarFile`

File: lib/pages/chat_details/chat_details_edit_view.dart [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+        focusNode: controller.groupNameFocusNode,
+        decoration: InputDecoration(
+          border: OutlineInputBorder(
+            borderSide: BorderSide(color: Theme.of(context).colorScheme.shadow),
+          ),
+          labelText: L10n.of(context)!.widgetName,
+          labelStyle: Theme.of(context).textTheme.bodySmall?.copyWith(
+                fontSize: 16,
+                letterSpacing: 0.4,
+                color: Theme.of(context).colorScheme.onSurface,
```
Comment (sherlockvn) [inline lib/pages/chat_details/chat_details_edit_view.dart:307]: move to style

**Comment by sherlockvn** [comment]
please update the demo and fix the error CI error, because i can't test the link

File: lib/domain/usecase/room/update_group_chat_interactor.dart (hunk 1 above)
Comment (drminh2807) [inline lib/domain/usecase/room/update_group_chat_interactor.dart:39]: you can use `Future.wait` to do all works in parallel

**Comment by sherlockvn** [comment]
image

Can you rebase main and check again ? because it's not fit

File: ios/Runner.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -83,10 +83,10 @@
 		97C146FC1CF9000F007C117D /* Main.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = 97C146FA1CF9000F007C117D /* Main.storyboard */; };
 		97C146FE1CF9000F007C117D /* Assets.xcassets in Resources */ = {isa = PBXBuildFile; fileRef = 97C146FD1CF9000F007C117D /* Assets.xcassets */; };
 		97C147011CF9000F007C117D /* LaunchScreen.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = 97C146FF1CF9000F007C117D /* LaunchScreen.storyboard */; };
+		B4B836C5699184AF03FB2652 /* Pods_Twake_Share.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 186FE0CC4065109DD7185CA1 /* Pods_Twake_Share.framework */; };
```
Comment (drminh2807) [inline ios/Runner.xcodeproj/project.pbxproj:86]: this PR should not touch this file

File: lib/domain/usecase/room/update_group_chat_interactor.dart (hunk 1 above)
Comment (drminh2807) [reply lib/domain/usecase/room/update_group_chat_interactor.dart:70]: not fixed yet

Comment (nqhhdev) [reply lib/domain/usecase/room/update_group_chat_interactor.dart:70]: If works in parallel, how to handle error?

Comment (drminh2807) [reply lib/domain/usecase/room/update_group_chat_interactor.dart:70]: add onError for each future, so you can override default error

File: assets/l10n/intl_en.arb [hunk 9]
```diff
@@ -2956,5 +2956,14 @@
...
+  "sharedMediaAndFiles": "Shared media and files",
+  "groupInformation": "Group information",
+  "linkInvite": "Link invite",
+  "noDescription": "No description",
+  "description": "Description",
+  "groupName": "Group name",
+  "descriptionHelper": "You can provide an optional description for your group.",
+  "errorUploadingGroupName": "Error uploading group name.",
+  "errorUploadingGroupDescription": "Error uploading group description.",
+  "errorUploadingGroupAvatar": "Error uploading group avatar."
```
Comment (hoangdat) [inline assets/l10n/intl_en.arb:2968]: Please ask designer for the common error. But we need to show the permission error in the case of no permission. (dont use the same error message for all)

File: lib/domain/usecase/room/update_group_chat_interactor.dart (hunk 1 above)
Comment (nqhhdev) [reply lib/domain/usecase/room/update_group_chat_interactor.dart:70]: I need to identify a specific error, not a general error. This is the reason why I have to make funcs synchronous.

File: lib/config/go_routes/go_router.dart [hunk 10]
```diff
@@ -438,50 +435,6 @@ abstract class AppRoutes {
...
-                      path: 'emotes/:state_key',
-                      pageBuilder: (context, state) => defaultPageBuilder(
-                        context,
-                        const EmotesSettings(),
-                      ),
-                      redirect: loggedOutRedirect,
-                    ),
-                  ],
-                  redirect: loggedOutRedirect,
-                ),
```
Comment (hoangdat) [inline lib/config/go_routes/go_router.dart:484]: why delete it?

File: lib/domain/usecase/room/update_group_chat_interactor.dart [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+import 'package:fluffychat/domain/app_state/room/update_group_chat_failure.dart';
+import 'package:fluffychat/domain/app_state/room/update_group_chat_loading.dart';
+import 'package:fluffychat/domain/app_state/room/update_group_chat_success.dart';
+import 'package:matrix/matrix.dart';
+
+class UpdateGroupChatInteractor {
+  Stream<Either<Failure, Success>> execute({
+    required Room room,
+    required Client client,
+    Uri? avatarUrl,
```
Comment (hoangdat) [inline lib/domain/usecase/room/update_group_chat_interactor.dart:13]: You don't upload avatar inside this interactor?

File: lib/pages/chat/chat.dart [hunk 12]
```diff
@@ -126,6 +125,9 @@ class ChatController extends State<Chat>
           EventTypes.Sticker,
           EventTypes.Encrypted,
           EventTypes.CallInvite,
+          EventTypes.RoomAvatar,
+          EventTypes.RoomName,
+          EventTypes.RoomTopic,
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:130]: maybe side effect for the case of creating chat?

File: lib/config/go_routes/go_router.dart (hunk 10 above)
Comment (nqhhdev) [reply lib/config/go_routes/go_router.dart:484]: Move Chat details in go_router to inner router inside chat

File: lib/pages/chat_details/chat_details_edit_context_menu_actions.dart [hunk 13]
```diff
@@ -0,0 +1,25 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+
+enum ChatDetailsEditContextMenuActions {
```
Comment (hoangdat) [inline lib/pages/chat_details/chat_details_edit_context_menu_actions.dart:4]:
```suggestion
enum EditChatAvatarContextMenuActions {

```

File: lib/domain/usecase/room/update_group_chat_interactor.dart (hunk 11 above)
Comment (nqhhdev) [reply lib/domain/usecase/room/update_group_chat_interactor.dart:13]: I use it inside chat, same logic update profile.

File: lib/pages/chat_details/chat_details_navigator.dart [hunk 14]
```diff
@@ -0,0 +1,50 @@
+import 'package:fluffychat/pages/chat_details/chat_details.dart';
+import 'package:fluffychat/pages/chat_details/chat_details_edit.dart';
+import 'package:flutter/material.dart';
+
+import 'package:fluffychat/presentation/model/presentation_contact.dart';
+
+class ChatDetailsRoutes {
+  static const String chatDetails = '/groupChatDetails';
+  static const String chatDetailsEdit = '/groupChatDetailsEdit';
```
Comment (hoangdat) [inline lib/pages/chat_details/chat_details_navigator.dart:9]: why not 
```suggestion
  static const String chatDetailsEdit = '/groupChatDetails/edit';
```

File: lib/pages/chat_details/chat_details_edit.dart [hunk 15]
```diff
@@ -0,0 +1,490 @@
...
+      );
+    } else {
+      _uploadGroupInfo(
+        displayName: groupNameTextEditingController.text,
+        description: descriptionTextEditingController.text,
+      );
+    }
+  }
+
+  void _uploadGroupInfo({
```
Comment (hoangdat) [inline lib/pages/chat_details/chat_details_edit.dart:318]:
```suggestion
  void _updateGroupInfo({
```

File: lib/pages/chat/chat.dart (hunk 12 above)
Comment (hoangdat) [reply lib/pages/chat/chat.dart:130]: wrong in the view of group creator. 

image

File: lib/pages/chat_details/chat_details.dart [hunk 16]
```diff
@@ -577,33 +267,38 @@ class ChatDetailsController extends State<ChatDetails>
...
+      context,
+      L10n.of(context)!.copiedToClipboard,
+    );
+  }
+
   @override
   Widget build(BuildContext context) {
     members ??=
         Matrix.of(context).client.getRoomById(roomId!)!.getParticipants();
     return SizedBox(
```
Comment (sherlockvn) [inline lib/pages/chat_details/chat_details.dart:300]: Put it in initState

File: lib/pages/chat_details/chat_details.dart [hunk 17]
```diff
@@ -577,33 +267,38 @@ class ChatDetailsController extends State<ChatDetails>
...
+  void onTapEditButton() {
+    Navigator.pushNamed(context, ChatDetailsRoutes.chatDetailsEdit);
+  }
+
+  void onTapInviteLink(BuildContext context, String inviteLink) async {
+    await Clipboard.setData(ClipboardData(text: inviteLink));
+    TwakeSnackBar.show(
+      context,
+      L10n.of(context)!.copiedToClipboard,
+    );
```
Comment (sherlockvn) [inline lib/pages/chat_details/chat_details.dart:293]: use `Clipboard.instance`, hide Clipboard from flutter

File: lib/pages/chat_details/chat_details_edit_view.dart [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+                              menuController.isOpen
+                                  ? menuController.close()
+                                  : menuController.open(),
+                            },
+                            style: ButtonStyle(
+                              shape: MaterialStateProperty.all(
+                                const CircleBorder(),
+                              ),
+                              padding: MaterialStateProperty.all(
+                                const EdgeInsets.all(8.0),
```
Comment (sherlockvn) [inline lib/pages/chat_details/chat_details_edit_view.dart:132]: style

**Comment by sherlockvn** [comment]
Demo mobile also

File: lib/domain/app_state/room/update_group_chat_failure.dart [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+class UpdateGroupDescriptionFailure extends Failure {
+  final dynamic exception;
+
+  const UpdateGroupDescriptionFailure({
+    this.exception,
+  }) : super();
+
+  @override
+  List<Object?> get props => [];
```
Comment (sherlockvn) [inline lib/domain/app_state/room/update_group_chat_failure.dart:54]: missing exception

File: lib/pages/chat_details/participant_list_item.dart [hunk 20]
```diff
@@ -20,7 +21,7 @@ class ParticipantListItem extends StatelessWidget {
       Membership.invite: L10n.of(context)!.invited,
       Membership.leave: L10n.of(context)!.leftTheChat,
     };
-    final permissionBatch = user.powerLevel == 100
+    final permissionBatch = user.isLevelAdmin
```
Comment (sherlockvn) [inline lib/pages/chat_details/participant_list_item.dart:24]: power level can be bigger than 100, don't change it at this time

File: lib/domain/usecase/room/update_group_chat_interactor.dart [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+            'url': isDeleteAvatar ? null : avatarUrl.toString(),
+          },
+        );
+        yield Right(UpdateAvatarGroupChatSuccess(isDeleteAvatar));
+      } on MatrixException catch (error) {
+        Logs().e(
+          'UpdateGroupChatInteractor::_updateAvatarForRoom(): ErrorCode ${error.errcode} - ErrorMessage ${error.errorMessage}',
+        );
+        yield Left(
+          UpdateGroupChatNoPermissionFailure(error),
```
Comment (hoangdat) [inline lib/domain/usecase/room/update_group_chat_interactor.dart:36]:
```suggestion
          UpdateGroupChatFailure(error),
```

File: lib/domain/usecase/room/update_group_chat_interactor.dart [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+              'name': displayName,
+            },
+          );
+          yield const Right(UpdateDisplayNameGroupChatSuccess());
+        } on MatrixException catch (error) {
+          Logs().e(
+            'UpdateGroupChatInteractor::_updateDisplayNameForRoom(): ErrorCode ${error.errcode} - ErrorMessage ${error.errorMessage}',
+          );
+          yield Left(
+            UpdateGroupChatNoPermissionFailure(error),
```
Comment (hoangdat) [inline lib/domain/usecase/room/update_group_chat_interactor.dart:62]:
```suggestion
            UpdateGroupChatFailure(error),
```

File: lib/domain/usecase/room/update_group_chat_interactor.dart [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+            'topic': description,
+          },
+        );
+        yield const Right(UpdateDescriptionGroupChatSuccess());
+      } on MatrixException catch (error) {
+        Logs().e(
+          'UpdateGroupChatInteractor::_updateDescriptionForRoom(): ErrorCode ${error.errcode} - ErrorMessage ${error.errorMessage}',
+        );
+        yield Left(
+          UpdateGroupChatNoPermissionFailure(error),
```
Comment (hoangdat) [inline lib/domain/usecase/room/update_group_chat_interactor.dart:93]:
```suggestion
          UpdateGroupChatFailure(error),
```

File: lib/pages/chat/chat.dart [hunk 24]
```diff
@@ -365,6 +364,9 @@ class ChatController extends State<Chat>
     loadTimelineFuture = _getTimeline();
     try {
       await loadTimelineFuture;
+      if ((timeline?.events.length ?? 0) < _loadHistoryCount) {
+        requestFuture();
+      }
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:369]: remove it (out of scope)

File: lib/pages/chat_details/chat_details_page_view/chat_details_members_page.dart [hunk 25]
```diff
@@ -26,14 +26,16 @@ class ChatDetailsMembersPage extends StatelessWidget {
   Widget build(BuildContext context) {
     return Column(
       children: [
-        _listMemberInfoMobileAndTablet(context),
         Expanded(
           child: ListView.builder(
             shrinkWrap: true,
-            itemCount: members.length + (canRequestMoreMembers ? 1 : 0),
+            itemCount: members.length + (canRequestMoreMembers ? 1 : 0) + 1,
```
Comment (hoangdat) [inline lib/pages/chat_details/chat_details_page_view/chat_details_members_page.dart:32]: `addMemberItemCount`

File: lib/pages/chat_details/chat_details_page_view/chat_details_members_page.dart [hunk 26]
```diff
@@ -26,14 +26,16 @@ class ChatDetailsMembersPage extends StatelessWidget {
...
             itemBuilder: (BuildContext context, int index) {
-              if (index < members.length) {
-                return ParticipantListItem(members[index]);
+              if (index == 0) {
+                return _listMemberInfoMobileAndTablet(context);
+              }
+              if (index - 1 < members.length) {
```
Comment (hoangdat) [inline lib/pages/chat_details/chat_details_page_view/chat_details_members_page.dart:37]:
```suggestion
              if (index - addMemberItemCount < members.length) {
```

File: lib/pages/chat_details/chat_details_page_view/chat_details_members_page.dart [hunk 27]
```diff
@@ -26,14 +26,16 @@ class ChatDetailsMembersPage extends StatelessWidget {
...
+                return ParticipantListItem(members[index - 1]);
```
Comment (hoangdat) [inline lib/pages/chat_details/chat_details_page_view/chat_details_members_page.dart:38]:
```suggestion
                return ParticipantListItem(members[index - addMemberItemCount]);
```

File: lib/presentation/extensions/room/user_extension.dart [hunk 28]
```diff
@@ -0,0 +1,7 @@
+import 'package:matrix/matrix.dart';
+
+extension UserExtension on User {
+  static const int levelAdmin = 100;
+
+  bool get isLevelAdmin => powerLevel == levelAdmin;
+}
```
Comment (hoangdat) [inline lib/presentation/extensions/room/user_extension.dart:7]: remove

File: lib/pages/chat_details/participant_list_item.dart (hunk 20 above)
Comment (hoangdat) [inline lib/pages/chat_details/participant_list_item.dart:24]: remove

**Review by sherlockvn — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**